### PR TITLE
Add log_image to wandb, cometml, and LoggerDestination

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?

Adds a `log_image` virtual method to `LoggerDestination` and implements `log_image` for `CometMLLogger` and `WandBLogger`. Also adds unit tests.

### WandBLogger.log_image

```python
import wandb
wandb.init()
import numpy as np
import torch
from composer.loggers import WandBLogger

wbl = WandBLogger()
wbl.log_images(images=[np.random.rand(3, 32,32), torch.rand(1, 32, 32)], name='my_cool_images', channels_last=False)
```
[results](https://wandb.ai/mosaic-ml/evan-composer/runs/3hd7yk48)

### CometMLLogger

```python
import numpy as np
import os
from composer.loggers import CometMLLogger
import torch

os.environ['COMET_API_KEY'] = '...' # redacted.
cl = CometMLLogger()

cl.log_images(images=[np.random.rand(3, 32,32), torch.rand(1, 32, 32)], name='my_cool_images', channels_last=False)
```
[results](https://www.comet.com/mosaic-ml/general/81a2ea57568646b0ba67d280a4503e6d?experiment-tab=images)

# What issue(s) does this change relate to?

fixes CO-1152
fixes CO-1153
fixes CO-1154

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/loggers/cometml_logger.py [hunk 1]
```diff
@@ -103,7 +106,42 @@ def log_hyperparameters(self, hyperparameters: Dict[str, Any]):
...
+    if np.any(np.equal(image.shape, 0)):
+        raise ValueError(f'Got an image (shape {image.shape}) with at least one dimension being 0! ')
+
+    if image.ndim > 3:
+        raise ValueError('Input image must be 1, 2, or 3 dimensions, but instead'
+                         f' got {image.ndim} dims at shape: {image.shape}'
+                         f' Your input image was interpreted as a batch of {image.ndim}-dimensional'
+                         f' images because you either specified a {image.ndim + 1}D image'
+                         f' or a list of {image.ndim}D images.'
+                         ' Please specify either a 4D image of a list of 3D images')
```
Comment (mvpatel2000) [inline composer/loggers/cometml_logger.py:145]: does this need a textwrap.dedent?

File: composer/loggers/wandb_logger.py [hunk 2]
```diff
@@ -267,3 +292,38 @@ def post_close(self) -> None:
...
+    # putting channels last.
+    if image.ndim == 1:
+        image = image[:, None, None]
+        channels_last = True
+    if image.ndim == 2:
+        image = image[:, :, None]
+        channels_last = True
+
+    if image.ndim != 3:
+        raise ValueError('Input image must be 3 dimensions, but instead'
```
Comment (mvpatel2000) [inline composer/loggers/wandb_logger.py:319]: same as above

**Review by mvpatel2000 — COMMENTED** [review]
Feel free to rerequest review once tests are passing

File: tests/loggers/test_cometml_logger.py [hunk 3]
```diff
@@ -1,18 +1,89 @@
...
+    from composer.loggers import CometMLLogger
+
+    # Set offline directory.
+    os.environ['COMET_OFFLINE_DIRECTORY'] = comet_offline_directory
+
+    comet_logger = CometMLLogger()
+    return comet_logger
+
+
+@pytest.mark.parametrize('images,channels_last', [(torch.rand(32), False), (torch.rand(32, 32), False),
```
Comment (dakinggg) [inline tests/loggers/test_cometml_logger.py:40]: when would you have a 1D image?

File: tests/loggers/test_wandb_logger.py [hunk 4]
```diff
@@ -23,6 +26,54 @@
...
+
+@pytest.mark.parametrize(
+    'images,channels_last',
+    [
+        (torch.rand(32), False),
+        (torch.rand(32, 0), False),  # Has zero in dimension.
+        (torch.rand(4, 4, 8, 32, 32), False),  # > 4 dim.
+        ([torch.rand(4, 32, 32, 3)], True),
+    ])  # sequence > 3 dim.
+def test_comet_ml_log_image_errors_out(comet_logger, images, channels_last):
```
Comment (dakinggg) [inline tests/loggers/test_wandb_logger.py]: this should probably be wandb logger?

**Comment by dakinggg** [comment]
Could you also confirm that you have manually tested that valid looking images actually get piped through to wandb/comet?

File: composer/loggers/wandb_logger.py [hunk 5]
```diff
@@ -267,3 +295,38 @@ def post_close(self) -> None:
...
+        raise ValueError(f'Got an image (shape {image.shape}) with at least one dimension being 0! ')
+
+    # Squeeze any singleton dimensions and then add them back in if image dimension
+    # less than 3.
+    image = image.squeeze()
+
+    # Add in length-one dimensions to get back up to 3
+    # putting channels last.
+    if image.ndim == 1:
+        image = image[:, None, None]
```
Comment (dakinggg) [inline composer/loggers/wandb_logger.py:315]: prefer using `np.newaxis` or `np.expand_dims` i think?

File: tests/loggers/test_cometml_logger.py (hunk 3 above)
Comment (eracah) [reply tests/loggers/test_cometml_logger.py:40]: Visualizing a 1D heatmap of a representation vector or the output of a softmax.

File: composer/loggers/logger.py [hunk 6]
```diff
@@ -66,6 +67,36 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
...
+                   channels_last: bool = False,
+                   step: Optional[int] = None,
+                   masks: Optional[Dict[str, Union[np.ndarray, torch.Tensor]]] = None,
+                   segmentation_class_labels: Optional[Dict[int, str]] = None):
+        """Log images. Logs any images from tensors or arrays
+
+        Args:
+            images (Dict[str,PIL.Image | np.ndarray | torch.Tensor ]): Dictionary mapping
+                image(s)' names (str) to an image of array of images.
+            name (str): The name of the image(s). (Default: ``'Images'``)
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/loggers/logger.py:82]: No it's a string

File: tests/loggers/test_cometml_logger.py (hunk 3 above)
Comment (dakinggg) [reply tests/loggers/test_cometml_logger.py:40]: :+1: thanks

File: composer/loggers/logger.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/loggers/logger.py:82]: oh, duh :)

File: composer/loggers/cometml_logger.py (hunk 1 above)
Comment (eracah) [reply composer/loggers/cometml_logger.py]: done

File: composer/loggers/wandb_logger.py (hunk 2 above)
Comment (eracah) [reply composer/loggers/wandb_logger.py]: done

File: composer/loggers/logger.py [hunk 7]
```diff
@@ -66,6 +67,36 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
...
+                   images: Union[np.ndarray, torch.Tensor, Sequence[Union[np.ndarray, torch.Tensor]]],
+                   name: str = 'Images',
+                   channels_last: bool = False,
+                   step: Optional[int] = None,
+                   masks: Optional[Dict[str, Union[np.ndarray, torch.Tensor]]] = None,
+                   segmentation_class_labels: Optional[Dict[int, str]] = None):
+        """Log images. Logs any images from tensors or arrays
+
+        Args:
+            images (Dict[str,PIL.Image | np.ndarray | torch.Tensor ]): Dictionary mapping
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/loggers/logger.py]: fixed

File: composer/loggers/logger.py [hunk 8]
```diff
@@ -66,6 +67,36 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
...
+                image(s)' names (str) to an image of array of images.
+            name (str): The name of the image(s). (Default: ``'Images'``)
+            channels_last (bool): Whether the channel dimension is first or last.
+                (Default: ``'False``)
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/loggers/logger.py]: fixed

File: composer/loggers/logger.py [hunk 9]
```diff
@@ -66,6 +67,36 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
...
+            step (Optional[int], optional): The current step or batch of training at the
+                time of logging. Defaults to None. If not specified the specific
+                LoggerDestination implementation will choose a step (usually a running
+                counter).
+            masks (Dict[str, PIL.Image | np.ndarray | torch.Tensor]): For segmentation inputs.
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/loggers/logger.py]: it will be a dict, but deleted for now and saving for a future PR

File: composer/loggers/logger.py [hunk 10]
```diff
@@ -66,6 +67,36 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
...
+                Dictionary mapping string to a dictionary specifying a 2D mask array
+            segmentation_class_labels (Dict[int, str]): For segmentation inputs. A class
+                label to name dictionary.
+
+        """
+        if not step:
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/loggers/logger.py]: good catch. fixed

File: composer/loggers/cometml_logger.py [hunk 11]
```diff
@@ -103,7 +106,42 @@ def log_hyperparameters(self, hyperparameters: Dict[str, Any]):
...
+                   segmentation_class_labels: Optional[Dict[int, str]] = None):
+
+        if self._enabled:
+            if not isinstance(images, Sequence) and images.ndim <= 3:
+                images = [images]
+
+            image_channels = 'last' if channels_last else 'first'
+            for image in images:
+                comet_image = _convert_to_comet_image(image)
+                assert self.experiment is not None
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/loggers/cometml_logger.py]: fixed

File: tests/loggers/test_wandb_logger.py (hunk 4 above)
Comment (eracah) [reply tests/loggers/test_wandb_logger.py]: fixed

**Comment by eracah** [comment]
> Could you also confirm that you have manually tested that valid looking images actually get piped through to wandb/comet?

@dakinggg, okay added manual tests

**Review by dakinggg — APPROVED** [review]
LGTM, thanks!

**Review by mvpatel2000 — APPROVED** [review]
LGTM. Leaning on Daniel's review here since I'm less familiar w this

**Comment by eracah** [comment]
Hey @bandish-shah, everything passes here, but Jenkins is for some reason failing. Can you PTAL?
